### ISSUE TITLE
Encoder-Decoder Example not working

### ISSUE THREAD

**Issue by jsuit** [Element-Research/rnn#213]
enc:backward(encInSeq, zeroTensor), line 77, gives error of inconsistent tensor size. I'm having a hard time figuring out this example, so I'm not having any luck figuring out how to fix it.

Also, why is opt.seqLen = 2, and not 3?

**Comment by frugs** [comment]
the zeroTensor should have dimensions 2, 6
`local zeroTensor = torch.Tensor(opt.seqLen, opt.vocabSize):zero()`

**Comment by frugs** [comment]
I believe @jsuit opt.seqLen is 2 because the length of the defined input sequences (and expected output sequence) is 2 - see lines 135, 137, and 140

**Comment by jsuit** [comment]
Why are we using a zeroTensor? Shouldn't it be the prevGradInput?

**Comment by frugs** [comment]
I believe it's because we don't actually care about the output of the encoder (only the hidden state of it after it's finished encoding), so we backpropagate the gradient of the output as as zero tensor so that it doesn't have any effects on the update of the weights and biases of the encoder LSTM.

**Comment by frugs** [comment]
I also agree that opt.seqLen should be 3 rather than 2.

**Comment by frugs** [comment]
The reason the zeroTensor must have dimensions (2, 6) is because this tensor represents the gradient of the output of the encoder, and therefore must have the same dimensions as the actual output of the encoder. Thus, the line should read as follows:

`local zeroTensor = torch.Tensor(encOut):zero()`

**Comment by frugs** [comment]
fixed in PR 216: https://github.com/Element-Research/rnn/pull/216

**Comment by nicholas-leonard** [comment]
Thanks again @frugs !

**Comment by jsuit** [comment]
Is the encoder trained at all? And why even do enc:backward call at all?

Jonathan Suit

**Comment by frugs** [comment]
Apologies, I hadn't fully understood how the code here worked when I answered the question, and it seems like the encoder-decoder design here is slightly different to the one I am familiar with, in that the output of the encoder is actually fed into the decoder, so please ignore my previous answer.

My understanding at the moment is that the encoder is trained - just the gradients we use for training are not passed as a parameter via the enc:backward call, but copied into encLSTM directly via the backwardConnect() function. It is still necessary to call enc:backward in order to initiate the back propagation, however.
